File: ios/Classes/FlutterMediapipeChatPlugin.swift
import Flutter
import UIKit

@available(iOS 13.0, *)
public class FlutterMediapipeChatPlugin: NSObject, FlutterPlugin {
    private var inferenceController: InferenceController?
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "flutter_mediapipe_chat/methods", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "flutter_mediapipe_chat/events", binaryMessenger: registrar.messenger())
        let instance = FlutterMediapipeChatPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    @MainActor 
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadModel":
            guard let arguments = call.arguments as? [String: Any],
                  let config = try? ModelConfig(dictionary: arguments) else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid model configuration", details: nil))
                return
            }
            do {
                inferenceController = try InferenceController(config: config)
                result(true)
            } catch {
                result(FlutterError(code: "MODEL_LOAD_ERROR", message: error.localizedDescription, details: nil))
            }
        case "generateResponse":
            guard let arguments = call.arguments as? [String: Any],
                  let prompt = arguments["prompt"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Prompt is required", details: nil))
                return
            }
            do {
                let response = try inferenceController?.generateResponse(prompt: prompt)
                result(response)
            } catch {
                result(FlutterError(code: "INFERENCE_ERROR", message: error.localizedDescription, details: nil))
            }
        case "generateResponseStream":
            guard let arguments = call.arguments as? [String: Any],
                  let prompt = arguments["prompt"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Prompt is required", details: nil))
                return
            }
            Task {
                do {
                    try await inferenceController?.generateResponseStream(prompt: prompt)
                    DispatchQueue.main.async {
                        result(nil)
                    }
                } catch {
                    DispatchQueue.main.async {
                        result(FlutterError(code: "INFERENCE_ERROR", message: error.localizedDescription, details: nil))
                    }
                }
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

@available(iOS 13.0, *)
extension FlutterMediapipeChatPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        inferenceController?.setEventSink(eventSink: events)
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        inferenceController?.clearEventSink()
        return nil
    }
}

File: android/src/main/kotlin/com/juandpt/flutter_mediapipe_chat/FlutterMediapipeChatPlugin.kt
package com.juandpt.flutter_mediapipe_chat

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class FlutterMediapipeChatPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var inferenceModel: InferenceModel? = null
    private lateinit var context: Context
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "flutter_mediapipe_chat/methods")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "flutter_mediapipe_chat/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "loadModel" -> {
                try {
                    val map = call.arguments as Map<*, *>
                    val config = ModelConfig.fromMap(map)
                    inferenceModel = InferenceModel.getInstance(context, config)
                    result.success(true)
                } catch (e: Exception) {
                    result.error("MODEL_LOAD_ERROR", e.localizedMessage, null)
                }
            }
            "generateResponse" -> {
                try {
                    val prompt = call.argument<String>("prompt") ?: throw IllegalArgumentException("Missing 'prompt'")
                    val response = inferenceModel?.generateResponse(prompt)
                    result.success(response)
                } catch (e: Exception) {
                    result.error("INFERENCE_ERROR", e.localizedMessage, null)
                }
            }
            "generateResponseAsync" -> {
                try {
                    val prompt = call.argument<String>("prompt") ?: throw IllegalArgumentException("Missing 'prompt'")
                    inferenceModel?.generateResponseAsync(prompt)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("INFERENCE_ERROR", e.localizedMessage, null)
                }
            }
            "close" -> {
                inferenceModel?.close()
                inferenceModel = null
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        inferenceModel?.let { model ->
            scope.launch {
                model.partialResults.collect { pair ->
                    eventSink?.success(pair.first)
                    if (pair.second) {
                        eventSink?.success(null)
                    }
                }
            }
            scope.launch {
                model.errors.collect { error ->
                    eventSink?.error("ERROR", error.message, null)
                }
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
}
